### PR TITLE
Implement placeholder-count validation and table/docstring diagnostics (LSP)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- This PR implements and finalizes the plan for placeholder-count validation and its companion diagnostics (LSP surface) including placeholder-count, data-table, and docstring validation. It adds an ExecPlan living document (docs/execplans/7-4-2-validate-placeholder-counts.md), expands diagnostics publication to cover new diagnostics, and includes unit and behavioural tests plus documentation updates.
- Adds a living ExecPlan detailing goals, constraints, progress, and staged plan: docs/execplans/7-4-2-validate-placeholder-counts.md.
- Extends diagnostics publication to publish placeholder-count and table/docstring diagnostics in addition to existing diagnostics.
- Introduces unit and behavioural tests to validate the new diagnostics.
- Updates documentation and repository roadmaps to reflect progress and planned gates.
- Artifacts and housekeeping updated to include new diagnostics-related files.
- No macro crates were modified; changes focus on the language server/diagnostics surface and tests.

## Changes
- Core diagnostics:
  - Added compute_signature_mismatch_diagnostics in crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs to validate that the number of placeholders in a step pattern matches the number of step arguments in the corresponding function signature. It uses lex_pattern() to extract placeholders and compares against non-fixture/non-datatable/docstring parameters.
  - Added helper logic: extract_placeholder_names(pattern), count_step_arguments(parameters, placeholder_names), and normalise_param_name(name) to support placeholder-based counting.
  - Added diagnostic builder build_placeholder_mismatch_diagnostic to emit a warning with a clear message and location.
- Table/docstring validation:
  - Added compute_table_docstring_mismatch_diagnostics to validate feature steps against Rust implementations on a per-step basis, including table and docstring expectations.
  - Implemented helpers to locate best matching implementation and build diagnostics for when a feature step provides a table/docstring that the Rust step does not expect (and vice versa).
- Diagnostics codes:
  - Added new codes in crates/rstest-bdd-server/src/handlers/diagnostics/mod.rs:
    - CODE_PLACEHOLDER_COUNT_MISMATCH = "placeholder-count-mismatch"
    - CODE_TABLE_EXPECTED = "table-expected"
    - CODE_TABLE_NOT_EXPECTED = "table-not-expected"
    - CODE_DOCSTRING_EXPECTED = "docstring-expected"
    - CODE_DOCSTRING_NOT_EXPECTED = "docstring-not-expected"
- Publication wiring:
  - Updated publish.rs to publish diagnostics for both placeholder-count mismatches and table/docstring mismatches in addition to existing diagnostics.
  - Extended rust diagnostics publishing to include compute_signature_mismatch_diagnostics alongside existing computations.
- Data model:
  - Extended CompiledStepDefinition to include parameters: Vec<IndexedStepParameter> so we can inspect function signatures and count step arguments accurately.
- Tests:
  - Expanded crates/rstest-bdd-server/tests/diagnostics.rs with tests exercising placeholder-count mismatch scenarios and basic table/docstring mismatch scenarios.
  - Added new behavioural tests: crates/rstest-bdd-server/tests/diagnostics_placeholder.rs and crates/rstest-bdd-server/tests/diagnostics_table_docstring.rs.
  - Added tests/support/diagnostics_helpers.rs to centralize test helpers.
- ExecPlan documentation:
  - Added ExecPlan document: docs/execplans/7-4-2-validate-placeholder-counts.md detailing goals, constraints, progress, and staged plan.
- Documentation & planning:
  - Updated docs/roadmap.md and docs/users-guide.md to reflect new diagnostics and learnings.
- Artifacts & housekeeping:
  - Updated scripts/rs-length-allowlist.txt to include new diagnostics-related files.
- No macro crates were modified; changes focus on the language server/diagnostics surface and tests.

## Rationale
- This work provides editor-friendly feedback for inconsistencies between Gherkin step patterns and Rust step signatures, including placeholder counts, and mismatches involving data tables and docstrings.
- Aligns with the existing diagnostic framework and keeps a single source of truth for placeholder extraction via lex_pattern().
- The changes support a phased plan with progressive validation (unit tests, behavioural tests, and documentation) while ensuring existing diagnostics remain intact.

## Plan of Work (Stage Overview)
- Stage A: Design (research only) - Design and prototyping done; no code changes required here.
- Stage B: Placeholder Count Validation - Completed:
  - Added diagnostic codes, implemented compute_signature_mismatch_diagnostics(), added helpers, wired into publication, and created unit tests.
- Stage C: Data Table & Docstring Validation - Completed:
  - Implemented compute_table_docstring_mismatch_diagnostics(), added necessary helpers, wired into publish flow, and added unit tests.
- Stage D: Behavioural Tests - Completed:
  - Added end-to-end tests for placeholder counts and table/docstring mismatches.
- Stage E: Documentation & Cleanup - Completed:
  - Updated users guide and roadmap; added ExecPlan; updated allowlist.

## Validation and Acceptance
- Quality gates: fmt, lint, and test pass.
- Unit tests cover placeholder counts and table/docstring validation.
- Behavioural tests verify end-to-end diagnostic publishing on save.
- Documentation updates planned in users guide and roadmap.

## Risks & Mitigations
- Placeholder counting semantics may diverge between macros and LSP indexing; resolved by using the same placeholder extraction and normalization rules, with tests covering edge cases.
- Adding new diagnostics should not affect existing diagnostics behavior; changes preserve current invariants.

## Artifacts
- New ExecPlan document: docs/execplans/7-4-2-validate-placeholder-counts.md

## Interfaces & Dependencies
- References to existing crates: rstest-bdd-server, rstest-bdd-patterns for pattern parsing and placeholders.
- No new external dependencies required.

## Related Work / Notes
- This ExecPlan is intended as a living document; updates will be reflected as work progresses and milestones are reached.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---
ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/5b3c0314-1f0d-4f74-894f-5d13a5f9ee96

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs](https://github.com/leynos/rstest-bdd/pull/396/files/de0a24a7b6b4a0c79ef3637990c8fc07c3ee8997#diff-f19f08bc80d0057cc6588a4ca8480055f3354195405f088395927d272855a2c0)

Comment on lines +410 to +429

~~~rust
fn build_docstring_expected_diagnostic(
    feature_index: &FeatureFileIndex,
    step: &IndexedStep,
) -> Diagnostic {
    let range = gherkin_span_to_lsp_range(&feature_index.source, step.span);

    Diagnostic {
        range,
        severity: Some(DiagnosticSeverity::WARNING),
        code: Some(lsp_types::NumberOrString::String(
            CODE_DOCSTRING_EXPECTED.to_owned(),
        )),
        code_description: None,
        source: Some(DIAGNOSTIC_SOURCE.to_owned()),
        message: "Step implementation expects a doc string but none is provided".to_owned(),
        related_information: None,
        tags: None,
        data: None,
    }
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=396&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-server%2Fsrc%2Fhandlers%2Fdiagnostics%2Fcompute.rs&method=)
The module contains 5 functions with similar structure: build_docstring_expected_diagnostic,build_docstring_not_expected_diagnostic,build_table_expected_diagnostic,build_table_not_expected_diagnostic and 1 more functions

File: crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs [hunk 1]
```diff
@@ -127,3 +213,233 @@ pub(super) fn step_type_to_attribute(step_type: gherkin::StepType) -> &'static s
...
+    step: &IndexedStep,
+) -> Option<Arc<CompiledStepDefinition>> {
+    state
+        .step_registry()
+        .steps_for_keyword(step.step_type)
+        .iter()
+        .filter(|compiled| compiled.regex.is_match(&step.text))
+        .max_by(|a, b| {
+            // Use literal length as a proxy for specificity (longer = more specific)
+            a.pattern.len().cmp(&b.pattern.len())
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- Exposing `handlers::diagnostics` and `handlers::diagnostics::compute` as public modules broadens the public surface area; consider re‑exporting only the specific diagnostic functions you intend external callers to use instead of making the entire modules `pub`.
- The placeholder/table/docstring mismatch scenarios are implemented both as unit tests in `compute.rs` and as behavioural tests in `tests/diagnostics.rs`; consider extracting shared helpers or consolidating some cases to reduce duplication and the risk of these suites drifting apart.
- In `find_best_matching_implementation` you currently use `pattern.len()` as a proxy for specificity; consider leveraging the existing `SpecificityScore` logic from `rstest-bdd-patterns` (or a similar scoring function) so that implementation selection for diagnostics stays aligned with the runtime matching semantics.

## Individual Comments

### Comment 1
<location> `crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs:264-245` </location>
<code_context>
+/// Uses `lex_pattern()` as the single source of truth for placeholder parsing.
+/// Returns `None` if the pattern cannot be lexed (malformed patterns are
+/// handled elsewhere and should not produce additional diagnostics here).
+fn extract_placeholder_names(pattern: &str) -> Option<HashSet<String>> {
+    let tokens = lex_pattern(pattern).ok()?;
+    let names = tokens
</code_context>

<issue_to_address>
**question:** Using a HashSet for placeholder names changes the meaning of “placeholder count” and may confuse users when the same placeholder appears multiple times.

Since `extract_placeholder_names` returns a `HashSet<String>`, `placeholder_count` is the number of *distinct* placeholder names, not total placeholder occurrences. For a pattern like `"I have {x} and {x}"`, this yields 1 instead of 2, which may not match what users perceive from the pattern text. If the intent is to count distinct identifiers for comparison with step arguments, consider updating the diagnostic wording (e.g., “distinct placeholder name(s)”), or alternatively track both total placeholder tokens and distinct names so the message aligns with user expectations.
</issue_to_address>

### Comment 2
<location> `crates/rstest-bdd-server/src/handlers/diagnostics/compute.rs:431-440` </location>
<code_context>
+///
+/// Returns the implementation with the highest specificity score if multiple
+/// match. Returns `None` if no implementation matches.
+fn find_best_matching_implementation(
+    state: &ServerState,
+    step: &IndexedStep,
+) -> Option<Arc<CompiledStepDefinition>> {
+    state
+        .step_registry()
+        .steps_for_keyword(step.step_type)
+        .iter()
+        .filter(|compiled| compiled.regex.is_match(&step.text))
+        .max_by(|a, b| {
+            // Use literal length as a proxy for specificity (longer = more specific)
+            a.pattern.len().cmp(&b.pattern.len())
+        })
+        .cloned()
</code_context>

<issue_to_address>
**suggestion:** Using raw pattern length as a proxy for specificity may pick unintuitive implementations when multiple patterns match.

Using `pattern.len()` for tie-breaking can rank a pattern with many placeholders/escapes as “more specific” than a shorter, more literal pattern, even though it’s semantically broader. Consider a heuristic based on literal content instead, e.g.:
- Count of literal vs placeholder tokens from `lex_pattern`, or
- Prefer fewer placeholders first, then break ties by literal length.
This should make selection more predictable when patterns overlap.

Suggested implementation:

```rust
fn implementation_specificity(compiled: &CompiledStepDefinition) -> (usize, usize) {
    // Heuristic:
    // 1. Prefer patterns with fewer placeholders (more literal structure).
    // 2. For the same number of placeholders, prefer patterns with more literal content.
    let tokens = lex_pattern(&compiled.pattern);

    let mut placeholder_count = 0usize;
    let mut literal_length = 0usize;

    for token in tokens {
        match token {
            PatternToken::Placeholder(_) => {
                placeholder_count += 1;
            }
            PatternToken::Literal(text) => {
                literal_length += text.len();
            }
        }
    }

    // Sort key: fewer placeholders first, then longer literal content.
    // We use (reverse_placeholder_count, literal_length) so that `max_by`
    // picks the most specific implementation.
    let reverse_placeholder_count = usize::MAX - placeholder_count;
    (reverse_placeholder_count, literal_length)
}

/// Find the best matching Rust implementation for a feature step.
///
/// Returns the implementation with the highest specificity score if multiple
/// match. Returns `None` if no implementation matches.
fn find_best_matching_implementation(
    state: &ServerState,
    step: &IndexedStep,
) -> Option<Arc<CompiledStepDefinition>> {
    state
        .step_registry()
        .steps_for_keyword(step.step_type)
        .iter()
        .filter(|compiled| compiled.regex.is_match(&step.text))
        .max_by(|a, b| {
            let a_score = implementation_specificity(a);
            let b_score = implementation_specificity(b);
            a_score.cmp(&b_score)
        })
        .cloned()
}

```

To make this compile and integrate correctly you will need to:

1. Import `lex_pattern` and the pattern token type used by step definitions at the top of this file, e.g.:
   - `use crate::pattern::lex_pattern;`
   - `use crate::pattern::PatternToken;`
   or the appropriate module paths for your project.
2. Ensure `PatternToken` has variants that distinguish placeholders from literal text (the example assumes something like:
   `enum PatternToken { Literal(String), Placeholder(PlaceholderKind), ... }`).
   Adjust the `match` in `implementation_specificity` to use the actual variant names and fields from your codebase.
3. If `lex_pattern` returns a slice or iterator instead of an owned collection, adjust the loop accordingly (e.g. iterate over references instead of owning `token`).
4. If your placeholder representation carries no data, change `PatternToken::Placeholder(_)` to `PatternToken::Placeholder`.
</issue_to_address>

### Comment 3
<location> `crates/rstest-bdd-server/src/handlers/diagnostics/mod.rs:108` </location>
<code_context>
+            "expected exactly 1 diagnostic, found {}",
+            diagnostics.len()
+        );
+        let diag = diagnostics.first().unwrap();
+        assert_eq!(
+            diag.code,
</code_context>

<issue_to_address>
**suggestion (review_instructions):** This new test helper uses `.unwrap()` instead of the preferred `.expect()`.

Even in test helpers, the guideline is to prefer `.expect()` over `.unwrap()` so failures are more self-describing.

Here you could do something like:
```rust
let diag = diagnostics
    .first()
    .expect("diagnostics.len() was asserted to be 1 above");
```
That keeps the intent clear while following the project convention.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Prefer `.expect()` over `.unwrap()`

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review

File: crates/rstest-bdd-server/tests/support/mod.rs [hunk 2]
```diff
@@ -3,7 +3,21 @@
 //! This module re-exports utilities from the crate's `test_support` module,
 //! providing a single source of truth for test infrastructure.
 
+// Allow unused imports since different test binaries use different subsets.
+#![allow(
+    unused_imports,
+    dead_code,
+    reason = "each test binary uses a different subset of helpers"
+)]
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd-server/tests/support/diagnostics_helpers/table_docstring.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+//! Helpers for table/docstring expectation mismatch diagnostic tests.
+
+#![allow(
+    dead_code,
+    reason = "only used by diagnostics_table_docstring test binary"
+)]
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/diagnostics_helpers/table_docstring.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd-server/src/handlers/diagnostics/placeholder.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        .iter()
+        .filter(|param| !param.is_datatable && !param.is_docstring)
+        .filter(|param| {
+            param
+                .name
+                .as_ref()
+                .is_some_and(|name| placeholder_names.contains(&normalize_param_name(name)))
+        })
+        .count()
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/src/handlers/diagnostics/placeholder.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd-server/tests/support/diagnostics_helpers/placeholder.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#[expect(dead_code, reason = "only used by diagnostics_placeholder test binary")]
+#[allow(unfulfilled_lint_expectations)]
+pub fn compute_placeholder_diagnostics(
+    state: &ServerState,
+    dir: &TempDir,
+    filename: impl AsRef<str>,
+) -> Vec<lsp_types::Diagnostic> {
+    let path = dir.path().join(filename.as_ref());
+    compute_signature_mismatch_diagnostics(state, &path)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/diagnostics_helpers/placeholder.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/rstest-bdd-server/tests/support/diagnostics_helpers/table_docstring.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+#[allow(unfulfilled_lint_expectations)]
+#[expect(clippy::expect_used, reason = "test helper uses expect for clarity")]
+pub fn compute_table_docstring_diagnostics(
+    state: &ServerState,
+    dir: &TempDir,
+    filename: impl AsRef<str>,
+) -> Vec<lsp_types::Diagnostic> {
+    let path = dir.path().join(filename.as_ref());
+    let feature_index = state.feature_index(&path).expect("feature index");
+    compute_table_docstring_mismatch_diagnostics(state, feature_index)
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/diagnostics_helpers/table_docstring.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/support/diagnostics_helpers/table_docstring.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
